Adds `hasLink` test for examples; update ~37 descriptions to pass (#474)
mapbox/mapbox-gl-js-docs: docs/pages/example/query-similar-features.md

- "import html from './query-similar-features.html';"
---

> In this example, a user can hover their cursor over any county in the United States, and the map highlights other U.S. counties with a matching name. The application also shows a popup with the name of the county at the cursor location, and shows an HTML overlay with summary information.

Author: Leaving first comma as-is because I think it improves readability for readers with English not as a first language (`the map highlights other U.S. counties with a matching name` is not a thing that `a user can hover their cursor over`).

> The example first uses [addSource](/mapbox-gl-js/api/map/#map#addsource) to connect data from a vector tileset, and then uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add two [`fill`](/mapbox-gl-js/style-spec/layers/#fill) layers using that source.

Reviewer: ```suggestion
The example first uses [`addSource`](/mapbox-gl-js/api/map/#map#addsource) to connect data from a vector tileset, and then uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add two [`fill`](/mapbox-gl-js/style-spec/layers/#fill) layers using that source.
```

> It uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to find other counties with the same name and highlights them using an overlay on the map.

Author: thank you! I've created https://github.com/mapbox/mapbox-gl-js-docs/issues/478 to track updating the example and here is the revised description – look ok?

> In this example, a user can hover their cursor over any county in the United States, and the map highlights other U.S. counties with a matching name, displays a popup with the name of the county at the cursor location, and displays an HTML overlay with summary information.
> 
> The example first uses [`addSource`](/mapbox-gl-js/api/map/#map#addsource) to connect data from a vector [tileset](https://docs.mapbox.com/help/glossary/tileset/). Then it uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add two [`fill` layers](/mapbox-gl-js/style-spec/layers/#fill) using that source, one for the a default display and one with all features highlighted with a color.
> 
> To add interactivity, it first uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to create a filter including only counties with a matching name, and finally uses [`setFilter`](/mapbox-gl-js/api/map/#map#setfilter) to add only matching features to the highlight layer.

> It uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to find other counties with the same name and highlights them using an overlay on the map.

Reviewer 2: >The example first uses addSource to connect data from a vector tileset. Then it uses addLayer to add two fill layers using that source, one for the a default display and one with all features highlighted with a color.

I think "connect" isn't the right choice here; "connect" the data to what? I would just say that it uses `addSource` to add the county polygons as a vector tileset.

The way this works is that we're adding two layers, one with transparent county polygons (the `rgba(0,0,0,0.1)` part means that the color is black at 10% opacity for the fill color and fill outline color) and one with filled county polygons. The filled county polygon layer has a filter on it that makes all the features in this layer not visible by default (it filters for the county name being an empty string which doesn't match anything). On `mousemove`, we use the mouse event to get the county feature that the mouse is currently over (`var feature = e.features[0];`). We use `querySourceFeatures` to get all the county features that match the current county name. We use that list of features to get the total population and create the overlay which is actually the white HTML box that's added to the map. Then we update the filter that's applied to the `counties-highlighted` layer (so instead of `['in', 'COUNTY', '']` it is now `['in' 'COUNTY', 'Washington']` for example). This then allows the matching counties to be visible, creating the illusion of a highlight feature. Finally, on `mouseleave`, we reset the `counties-highlighted` filter to the empty string and remove the overlay box.

> It uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to find other counties with the same name and highlights them using an overlay on the map.

Author: Thank you! I've made some improvements to both the code comments and the example description – I think we are good: c11fd627946d259e355a2f4d8adf6c98efd5ac08

> It uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses and [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to find other counties with the same name and highlights them using an overlay on the map.

Reviewer: ```suggestion
It uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to find other counties with the same name and highlights them using an overlay on the map.
```

> Then, it uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to gain [MapMouseEvent](/mapbox-gl-js/api/events/#mapmouseevent) information to use with [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to search for the county located at the cursor position, and other counties with the same name.

Reviewer: ```suggestion
It uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses and [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to find other counties with the same name and highlights them using an overlay on the map.
```
